### PR TITLE
[SYCL][Doc] Add "device-if" extension

### PR DESCRIPTION
**Pull request by gmlueck**
Add a new proposed extension specification that allows device code to
conditionally use features that are supported on only some devices.

### CODE REVIEW

File: sycl/doc/extensions/DeviceIf/device_if.asciidoc [hunk 1]
```diff
@@ -0,0 +1,391 @@
...
+In JIT mode, the goal is to generate a single SPIR-V module for the device
+code, which contains all the condition checks and lambda calls.  The online
+compiler evaluates the condition checks at online compilation time, and the
+generated native code contains only the body of the selected lambda.  The
+implementation can make use of specialization constants to represent each
+aspect: a specialization constant will be set to `true` if the device supports
+that aspect and to `false` if it does not.
+
+One challenge with this approach is that we need some way to guarantee that the
+online compiler will not raise a compilation error from an unselected lamdba
```
Comment (Pennycook) [inline sycl/doc/extensions/DeviceIf/device_if.asciidoc:307]:
```suggestion
online compiler will not raise a compilation error from an unselected lambda
```

File: sycl/doc/extensions/DeviceIf/device_if.asciidoc [hunk 2]
```diff
@@ -0,0 +1,391 @@
...
+call.  To illustrate, consider a call to `ext_oneapi_if_device_has()` which
+conditionally uses `sycl::half` and let's consider the case where the current
+device does not have this support.  We need to make sure that the online
+compiler does not raise a compilation error when attempting to compile the
+lambda body which uses `sycl::half`.  In such a case the SPIR-V will
+conceptually look like:
+
+```
+if (__builtin_spec_constant(/*SPIR-V ID for aspect::fp16*/)) call lamdba
```
Comment (Pennycook) [inline sycl/doc/extensions/DeviceIf/device_if.asciidoc:316]:
```suggestion
if (__builtin_spec_constant(/*SPIR-V ID for aspect::fp16*/)) call lambda
```

File: sycl/doc/extensions/DeviceIf/device_if.asciidoc [hunk 3]
```diff
@@ -0,0 +1,391 @@
...
+called at runtime, we don't want the online compiler to throw an exception when
+it tries to generate device code for the non-existent `sycl::half` type.  We
+therefore need some way to ensure that the online compiler discards any
+unselected lambda calls (or at least a way to ensure that the compiler doesn't
+throw an exception when compiling it).
+
+We propose introducing some SPIR-V extension that provides this guarantee.
+The extension would take a boolean expression of specialization constant values
+to compute an "if" condition.  The extension would then guarantee that the
+body of the "if" (the lamdba call) is discarded if the boolean expression is
```
Comment (Pennycook) [inline sycl/doc/extensions/DeviceIf/device_if.asciidoc:333]:
```suggestion
body of the "if" (the lambda call) is discarded if the boolean expression is
```

File: sycl/doc/extensions/DeviceIf/device_if.asciidoc [hunk 4]
```diff
@@ -0,0 +1,391 @@
...
+
+This extension provides a way for device code to query the device on which it
+is running in order to conditionally use features that may not be supported on
+all devices.  This is different from the existing `device::has()` function
+because the extension can be called from device code while `device::has()` can
+only be called from host code.
+
+The motivating use case for this extension is for developers who provide device
+side libraries.  To illustrate, consider an application developer (i.e. someone
+who is **not** developing a device side library) who wants to code a kernel
```
Comment (Pennycook) [inline sycl/doc/extensions/DeviceIf/device_if.asciidoc:43]: Nit: I think "device-side library" is easier to read here (and same comment for other usages of "device side library").

**Review by Pennycook — CHANGES_REQUESTED** [review]
Mostly LGTM, just spotted a few typos.

**Review by Pennycook — APPROVED** [review]

File: sycl/doc/extensions/DeviceIf/device_if.asciidoc (hunk 4 above)
Comment (gmlueck) [reply sycl/doc/extensions/DeviceIf/device_if.asciidoc]: Thanks, makes sense.

**Comment by keryell** [comment]
This reminds me what @brycelelbach presented today at GTC https://youtu.be/KhZvrF_w1ak?t=1607 on the new `if target` in `nvc++` replacing the CUDA `__CUDA_ARCH__` to avoid some ODR violation. Of course, since SYCL is pure C++, we do not want such a language extension like in `nvc++`...

**Pull request merged**